File: content/07-trees/validate_bst.js
class TreeNode {
    constructor(val = 0, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

function isValidBST(root) {
    function valid(node, min, max) {
        if (!node) return true;
        if (node.val <= min || node.val >= max) return false;
        return valid(node.left, min, node.val) && valid(node.right, node.val, max);
    }
    return valid(root, -Infinity, Infinity);
}

const assert = require('assert');
const r1 = new TreeNode(2, new TreeNode(1), new TreeNode(3));
assert.strictEqual(isValidBST(r1), true);
console.log("All tests passed!");

File: content/07-trees/validate_bst.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isValidBST(root: TreeNode) -> bool:
    def valid(node, left, right):
        if not node:
            return True
        if not (left < node.val < right):
            return False
        return valid(node.left, left, node.val) and valid(node.right, node.val, right)
    return valid(root, float("-inf"), float("inf"))

if __name__ == "__main__":
    r1 = TreeNode(2, TreeNode(1), TreeNode(3))
    assert isValidBST(r1) == True, "Test 1 Failed"
    print("All tests passed!")
